lua/toggleterm/terminal.lua

```diff
@@ -325,13 +327,31 @@ end
 ---@private
 ---Pass self as first parameter to callback
 function Terminal:__stdout()
-  if self.on_stdout then return function(...) self.on_stdout(self, ...) end end
+  if self.auto_scroll or self.on_stdout then
```

Comment:
```suggestion
  if self.on_stderr then
```
This check decides if there should be an on stderr callback at all it shouldn't include a check for autoscroll since that shouldn't be related to initializing the callback

Reply:
Umm.. Don't want to sound defensive, but why?
If there is no `on_stdout` and `auto_scroll` is disabled, why even bother creating a callback that will return false for both checks on every stdout?

Reply:
I'm not sure if I understand what you mean. The `__stderr` function returns a function which becomes the `on_stderr` or `nil`. So all it needs to care about is if a user specified an `on_stderr` this check has nothing to do with whether a user specified `auto_scroll`. Autoscroll doesn't decide if there is a callback here or not, only the presence of a user defined call back should be involved there.

`__on_stderr` itself is not itself the actual handler, it's just a convenience function to return a function or nil.

It is already called here
https://github.com/akinsho/toggleterm.nvim/blob/5ea823359b450bcf9a784f73734fd2d52e9de0ec/lua/toggleterm/terminal.lua#L354
*once* and so it returns the function that will be the `on_stderr` handler. It's not dissimilar from just doing `on_stderr = config.on_stderr and config.on_stderr or nil`

Reply:
Yeah.. I get it. It's a builder :)
My point is that if the user has not provided a custom `on_std(out|err)`, then the auto_scroll code would not run (the builder would return `nil`).
I understand that this could be unconditional and always return a callback for `on_std(out|err)`, but that would be quite silly if both `auto_scroll == false` and `self.on_std(out|err) == nil`. And since this callback is called quite often, I put in the check to to see if it even makes sense to build a callback

Reply:
I guess I'm genuinely not following why `auto_scroll` provides some efficiency by being in this conditional? I can't see how it factors into the determination of if there should be an `on_std{err|out}` callback at all?

It should just decide if it runs the `G` command or not, not do anything else.

> would be quite silly if both auto_scroll == false and self.on_std(out|err) == nil

I don't know why the first part i.e. `auto_scroll == false` is relevant to the second check? Am I missing something in this PR about auto_scroll being implicitly tied to there being an `on_std{err|out}`?

>  And since this callback is called quite often, I put in the check to to see if it even makes sense to build a callback

Specifically `__std{err|out}` are *not called often* only the functions they return are.

Reply:
Ah, right sorry that makes sense now I didn't realise that it implicitly is ensuring that a callback is returned if auto scroll is true, so that it works regardless of if the user specifies a custom callback. The only reason it's actually even in this builder was more for syntactic purposes, so wasn't even thinking about the fact that it could be used by something else as a callback.

I'd leave a small comment there, mainly because I'll probably forget that that's what it's doing at some point.

Reply:
What if I rename the generator function? The  `__std{err|out}` one to a name that conveys better what it is doing?
I was trying to minimize the footprint of my PR, but if you are ok with that I can propose it with my new push

Reply:
Yeah I don't mind a name change what did you have in mind?

Reply:
Pushed my suggestion: `__make_output_handler` to convey that it is "building" the handler.
I also unified the duplicated code.
I was going to separate the "including self" part into a separate function, but I think it is fair to say that that is part of "building" the handler